### PR TITLE
[IMP] *: remove jQuery.nested_sortable

### PR DESCRIPTION
**Pull request by res-odoo**
---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by rdeodoo** [comment]
@detrouxdev is working on that too in website :grimacing: 
He will have a look and approve on behalf of website :+1:

**Comment by rdeodoo** [comment]
Oh, and it's `task-3439226` not `Task ID: 3439226`, otherwise it sometimes creates wrong links pointing to random commits :grimacing: 
Just follow the guidelines :p (random example: `https://github.com/odoo/odoo/pull/90975/commits`)

**Review by rfr-odoo — APPROVED** [review]

File: addons/website/static/src/components/dialog/edit_menu.js [hunk 1]
```diff
@@ -122,40 +121,75 @@ export class EditMenuDialog extends Component {
...
+
     populate(map, menu) {
         map.set(menu.fields['id'], menu);
         for (const submenu of menu.children) {
             this.populate(map, submenu);
         }
     }
 
+    _isAllowedMove(current, elementSelector) {
+        const currentIsMegaMenu = current.element.dataset.isMegaMenu === "true"
```
Comment (detrouxdev) [inline addons/website/static/src/components/dialog/edit_menu.js:151]: Missing `;` (Unfortunately, the default eslint config is disabled on website, so this tends to happen)

File: addons/website/static/src/components/dialog/edit_menu.js [hunk 2]
```diff
@@ -112,8 +112,7 @@ export class EditMenuDialog extends Component {
 
         this.menuEditor = useRef('menu-editor');
 
-        this.state = useState({ rootMenu: {} });
-
+        this.state = useState({rootMenu: {}, flatMenus: new Map()});
```
Comment (detrouxdev) [inline addons/website/static/src/components/dialog/edit_menu.js:115]: I'm not sure I understand why this was necessary 😅 It looks like the state isn't used in the template, and every time we touch the map, we also change rootMenu (or one of its children) so a re-render is already triggered right?

File: addons/website/static/src/components/dialog/edit_menu.js [hunk 3]
```diff
@@ -122,40 +121,75 @@ export class EditMenuDialog extends Component {
...
+            previous = this.map.get(this._getMenuIdForElement(previous));
+            const index = parentMenu.children.findIndex((child) => child === previous);
+            parentMenu.children.splice(index + 1, 0, menu);
+        } else if (next) {
+            next = this.map.get(this._getMenuIdForElement(next));
+            const index = parentMenu.children.findIndex((child) => child === next);
+            parentMenu.children.splice(index, 0, menu);
+        } else {
+            parentMenu.children.push(menu);
+        }
```
Comment (detrouxdev) [inline addons/website/static/src/components/dialog/edit_menu.js:190]:
```suggestion
        if (previous) {
            previous = this.map.get(this._getMenuIdForElement(previous));
            const index = parentMenu.children.findIndex((child) => child === previous);
            parentMenu.children.splice(index + 1, 0, menu);
        } else {
            parentMenu.children.unshift(menu);
        }
```
If it doesn't have a previous, it means it's the first right? 😄 Whether it has a next element shouldn't matter I think.

**Review by detrouxdev — CHANGES_REQUESTED** [review]
Nice to see that nestedSortable is complete now 😄 A few remarks that are mostly suggestions. Thanks for the work! 👍🏻

**Comment by res-odoo** [comment]
> Nice to see that nestedSortable is complete now 😄 A few remarks that are mostly suggestions. Thanks for the work! 👍🏻

Hi @detrouxdev, I made all suggestions changes. There are pertinent 👍🏻

**Comment by detrouxdev** [comment]
@res-odoo Thanks for applying my suggestions! After a few more tests, I realised that you can drop a menu with child elements into another menu that already has a child element:
Screenshot 2023-09-29 at 12 30 57

I don't think this is only related to the website code, but if it is then I guess we should just check if a menu has children in the `_isAllowedMove` method?

**Comment by res-odoo** [comment]
> @res-odoo Thanks for applying my suggestions! After a few more tests, I realised that you can drop a menu with child elements into another menu that already has a child element: Screenshot 2023-09-29 at 12 30 57
> 
> I don't think this is only related to the website code, but if it is then I guess we should just check if a menu has children in the `_isAllowedMove` method?

You found a bug. To fix it, I made some changes in the `nestedSortable` hooks. Now the deepest check for the max level and the allowed move are both made inside a common method call when the `onMove` and the `onDrop` are called. In the case of an unallowed drop, it simply adds a `d-none` class to hide the drop zone from the end user.

**Review by detrouxdev — APPROVED** [review]
Thanks for fixing the bug! I actually think it also slightly improves the UX if a user tries to drag a mega menu as a child menu since there is no longer a dropzone 👍🏻

**Review by pparidans — APPROVED** [review]
Nice work :sunglasses: 

@robodoo r+ rebase-ff

**Pull request closed without merging**